### PR TITLE
Migrate solidity tests to end-to-end

### LINKED ISSUE #722: migrate test_erc20.py

### PR DESCRIPTION
**Pull request by ClementWalter**
Time spent on this PR: 0.5

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Ported Solmate's tests are run against the `starkware.starknet.testing.starknet.Starknet` starknet mock.

Resolves #722 #736

## What is the new behavior?

Run against any Starknet network.

Note: I had to pass failing test because katana currenly doesn't return REVERTED, see
https://github.com/dojoengine/dojo/issues/864

### CODE REVIEW

**Review by greged93 — APPROVED** [review]
lgtm!

File: tests/end_to_end/Solmate/test_erc20.py [hunk 1]
```diff
@@ -38,142 +34,160 @@ async def test_should_set_name_symbol_and_decimals(self, erc_20):
...
-                to_wallet.address,
-                TEST_SUPPLY,
-                caller_address=owner,
-            )
-            assert await erc_20.totalSupply() == TEST_SUPPLY
-            assert await erc_20.allowance(from_wallet.address, owner.address) == MAX_INT
-            assert await erc_20.balanceOf(from_wallet.address) == 0
-            assert await erc_20.balanceOf(others[1].address) == TEST_SUPPLY
-
+        @pytest.mark.xfail(
```
Comment (Eikix) [inline tests/end_to_end/Solmate/test_erc20.py:149]: did we create an issue to track the dojo issue?

Comment (ClementWalter) [reply tests/end_to_end/Solmate/test_erc20.py:149]: not really, I think that we should create an issue to track all `xfail` tests, like every week.
(Note: I've replaced some `mark.skip` by `mark.xfail` after re-reading the pytest doc; `xfail` and `xpass` are indeed to be used to keep the test running, not breaking the CI, but indicate that the intented behavior doesn't work)

Comment (Eikix) [reply tests/end_to_end/Solmate/test_erc20.py:149]: yeah maybe we should find a way for automatically check all the xfail and see if there are resolved or how we can resolve them

Comment (Eikix) [reply tests/end_to_end/Solmate/test_erc20.py:149]: a weekly telegram bot

**Review by greged93 — COMMENTED** [review]
Would suggest not adding more to this PR, it's getting hard to review

File: scripts/utils/starknet.py [hunk 2]
```diff
@@ -146,11 +150,12 @@ async def fund_address(
...
         response = requests.post(
             f"http://127.0.0.1:5050/mint",
-            json={"address": hex(address), "amount": amount},
+            json={"address": hex(address), "amount": int(amount)},
         )
         if response.status_code != 200:
             logger.error(f"Cannot mint token to {address}: {response.text}")
-        logger.info(f"{amount / 1e18} ETH minted to {hex(address)}")
+        else:
+            logger.info(f"{amount / 1e18} ETH minted to {hex(address)}")
```
[the remark this answers is not in the document]
Comment (ClementWalter) [reply scripts/utils/starknet.py:158]: not sure to follow, amount is declared as `amount: float` in the function signature

File: tests/end_to_end/PlainOpcodes/test_counter.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    class TestDec:
+        @pytest.mark.xfail(
+            os.environ.get("STARKNET_NETWORK", "katana") == "katana",
+            reason="https://github.com/dojoengine/dojo/issues/864",
+        )
+        async def test_should_raise_when_count_is_0(self, counter, owner):
+            with kakarot_error("count should be strictly greater than 0"):
+                await counter.dec(caller_eoa=owner)
+
+        async def test_should_decrease_count(self, counter, owner):
```
[the remark this answers is not in the document]
Comment (ClementWalter) [reply tests/end_to_end/PlainOpcodes/test_counter.py:38]: this is the responsibility of the `TestInc` to make sure that `.inc()` actually works. In a test for `.dec()`, one assumes that `.inc()` works

File: tests/end_to_end/PlainOpcodes/test_plain_opcodes.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+            When deploying a contract, the constructor bytecode is run but not
+            stored eventually,
+            """
+            deployed_bytecode = counter.bytecode[counter.bytecode.index(0xFE) + 1 :]
+            size = len(deployed_bytecode) if size is None else size
+            bytecode = await plain_opcodes.opcodeExtCodeCopy(offset=offset, size=size)
+            assert bytecode == deployed_bytecode[offset : offset + size]
+
+    class TestLog:
```
[the remark this answers is not in the document]
Comment (ClementWalter) [reply tests/end_to_end/PlainOpcodes/test_plain_opcodes.py:94]: I don't understand, we use the starknet event to build the corresponding EVM event in `parse_starknet_events`

File: scripts/utils/starknet.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (greged93) [reply scripts/utils/starknet.py:158]: ok my bad, because of line 153 I thought amount was a string

**Review by greged93 — APPROVED** [review]
lgtm!

**Pull request merged**